Adds Batch client connection and async RPC calls to the bitcoind client

Vib-UX opened the pull request:
# PR serves in relation to fixing issue [5041](https://github.com/lightningnetwork/lnd/issues/5041#issue-810611891)

With this PR we enable batch client connection and async RPC calls to the bitcoind client providing testcoverages for the same. This helps us in batching the required requests needed for speeding up operations by reducing round trips. 


## Pull Request Checklist
### Testing
* [ ]  Your PR passes all CI checks.
* [x]  Tests covering the positive and negative (error paths) are included.
* [ ]  Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
* [x]  The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
* [x]  Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).


📝 Please see our [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

guggero commented on chain/bitcoind_client.go:
```diff
@@ -150,6 +151,18 @@ func (c *BitcoindClient) GetBlockVerbose(
+// GetBestBlockHashAsync returns an instance of a type that can be used to get
+// the result of the RPC at some future time by invoking the Receive function on
+// the returned instance.
+func (c *BitcoindClient) GetBlockHashAsync(height int64) (rpcclient.FutureGetBlockHashResult, error) {
```
nit: line too long, should break before `height` parameter and then add a newline before the method body (since this then counts as a multi-line method declaration). See https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#code-spacing-and-formatting for general styling information, we try to adhere to those in the `btcsuite` projects as well (even though a lot of the existing code doesn't conform yet).

guggero commented on chain/bitcoind_client.go:
```diff
@@ -150,6 +151,18 @@ func (c *BitcoindClient) GetBlockVerbose(
+	return c.chainConn.batchClient.GetBlockHashAsync(height), nil
+}
+
+// func to send bulk request
```
nit: godoc comment must start with method name and should be full sentence.

guggero commented on chain/bitcoind_conn.go:
```diff
@@ -101,6 +101,9 @@ type BitcoindConn struct {
 	// client is the RPC client to the bitcoind node.
 	client *rpcclient.Client
 
+	// batch-Client is the RPC client to the bitcoind node
```
nit: a godoc comment should start with the exact variable name, so this should say `batchClient` instead of `batch-Client`.

bhandras replied:
nit: comments should always end with a punctuation.

guggero commented on chain/bitcoind_client.go:
```diff
@@ -150,6 +151,18 @@ func (c *BitcoindClient) GetBlockVerbose(
+}
+
+// func to send bulk request
+func (c *BitcoindClient) Send() error {
```
Should this method be called `SendAsyncQueue` perhaps? Otherwise in this context just "Send" doesn't really say what it's doing.

bhandras replied:
+1, it's worth thinking in less simplicity and more towards readability of the code, where more readable names tend to be not too long and not too short :)

guggero reviewed (commented):
Looks pretty good! Mostly style nits, functionality looks to there :+1:

bhandras commented on chain/bitcoind_conn.go:
```diff
@@ -101,7 +101,7 @@ type BitcoindConn struct {
 	client *rpcclient.Client
 
-	// batch-Client is the RPC client to the bitcoind node
+	// batchClient is the RPC client to the bitcoind node.
```
Please squash fixup commits, so that the pull request always consists of close to final state.

Vib-UX replied:
Sure

bhandras commented on chain/bitcoind_client.go:
```diff
@@ -381,16 +392,33 @@ func (c *BitcoindClient) RescanBlocks(
+	// request map will store rpcclient.FutureGetBlockHeaderVerboseResult for
+	// each blockHash.
+	request := make(
+		map[chainhash.Hash]rpcclient.FutureGetBlockHeaderVerboseResult, len(
```
nit: formatting

bhandras commented on chain/bitcoind_client.go:
```diff
@@ -381,16 +392,33 @@ func (c *BitcoindClient) RescanBlocks(
-		header, err := c.GetBlockHeaderVerbose(&hash)
+		headerRequest, err := c.GetBlockHeaderVerboseAsync(&hash)
+		if err != nil {
+			log.Warnf("Unable to receive header promise: %s", err)
```
nit: could log the hash too.

Vib-UX replied:
Yes

bhandras commented on chain/bitcoind_client.go:
```diff
@@ -381,16 +392,33 @@ func (c *BitcoindClient) RescanBlocks(
+
+	c.SendAsyncQueue()
+
+	for hash, headerRequest := range request {
```
nit: since `hash` is a local variable tracking the current element, it's worth adding `hash := hash` below this line such that any functions where a pointer to hash is passed don't end up tracking the last element in `requests`.

Vib-UX replied:
Thanks for the explanation, thats insightful!

bhandras commented on chain/bitcoind_client.go:
```diff
@@ -419,6 +419,7 @@ func (c *BitcoindClient) RescanBlocks(
 				hash, err)
 			continue
 		}
+
```
Squash with previous commit.

bhandras commented on chain/bitcoind_client.go:
```diff
@@ -962,17 +962,21 @@ func (c *BitcoindClient) FilterBlocks(
 
 	blockFilterer := NewBlockFilterer(c.chainConn.cfg.ChainParams, req)
 
+	storeHash := make([]*chainhash.Hash, len(req.Blocks))
```
The correct way to declare this slice would be `make([]*chainHash.Hash, 0, len(req.Blocks))` otherwise it'll have a predefined size and then we'll also append `len(req.Blocks)` elements to it.

bhandras commented on chain/bitcoind_client.go:
```diff
@@ -962,17 +962,21 @@ func (c *BitcoindClient) FilterBlocks(
 
+	storeHash := make([]*chainhash.Hash, len(req.Blocks))
+
+	for _, block := range req.Blocks {
```
Same scope bug as above. Here's a quick example of what's happening: https://go.dev/play/p/Q7OdkHdPa_g
If you run that example you'll see that we're storing the pointer to the same element every time.

Vib-UX replied:
Attached example was insightful,
https://medium.com/swlh/use-pointer-of-for-range-loop-variable-in-go-3d3481f7ffc9

Iterating variable with `range` re-assigns value to the same address location. Got all my doubts clear by reading the above blog, thanks for pointing it out.

bhandras commented on chain/bitcoind_conn.go:
```diff
@@ -600,6 +600,70 @@ func (c *BitcoindConn) GetBlock(hash *chainhash.Hash) (*wire.MsgBlock, error) {
+func (c *BitcoindConn) GetBlocksBatch(
+	hashes []*chainhash.Hash) ([]*wire.MsgBlock, error) {
+
+	batchRequest := make(map[*chainhash.Hash]rpcclient.FutureGetBlockResult,
```
nit: formatting.

bhandras commented on chain/bitcoind_conn.go:
```diff
@@ -600,6 +600,70 @@ func (c *BitcoindConn) GetBlock(hash *chainhash.Hash) (*wire.MsgBlock, error) {
+	batchRequest := make(map[*chainhash.Hash]rpcclient.FutureGetBlockResult,
+		len(hashes))
+
+	batchRawBlocks := make([]*wire.MsgBlock, 0, len(hashes))
```
nit: move declaration closer to usage.

bhandras commented on chain/bitcoind_conn.go:
```diff
@@ -600,6 +600,70 @@ func (c *BitcoindConn) GetBlock(hash *chainhash.Hash) (*wire.MsgBlock, error) {
+	}
+
+	return batchRawBlocks, nil
+
```
nit: extra nl

bhandras commented on chain/bitcoind_conn.go:
```diff
@@ -600,6 +600,70 @@ func (c *BitcoindConn) GetBlock(hash *chainhash.Hash) (*wire.MsgBlock, error) {
+		batchRequest[hash] = sendRequest
+	}
+
+	c.batchClient.Send()
```
Returned error not checked.

bhandras reviewed (commented):
Nice work @Vib-UX :zap: One general comment I can add is to please add some more detail to your commit messages. In general a good commit message is similar to what's describe in this doc:https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#model-git-commit-messages

Obviously it's not a requirement, but a good practice. At a minimum add some short description about what your commit changes, eg. `packagename: adding a new shiny func to achieve some goals.`

Looking forward for some test coverage as well :)

Vib-UX committed:
chain: adds batched async calls support for bitcoind client

This commit adds a batch client connection and async RPC
calls to the bitcoind client which will enable batching multiple
requests speeding up operations by reducing round trips.

bhandras commented on chain/bitcoind_client.go:
```diff
@@ -366,10 +393,31 @@ func (c *BitcoindClient) RescanBlocks(
+
+	// request map will store rpcclient.FutureGetBlockHeaderVerboseResult for
+	// each blockHash.
+	request := make(
```
nit: I think renaming this to `requests` is slightly better. Also fix formatting.

bhandras commented on chain/bitcoind_client.go:
```diff
@@ -366,10 +393,31 @@ func (c *BitcoindClient) RescanBlocks(
-		header, err := c.GetBlockHeaderVerbose(&hash)
+		headerRequest, err := c.GetBlockHeaderVerboseAsync(&hash)
+		if err != nil {
+			log.Warnf(
```
nit: formatting

bhandras commented on chain/bitcoind_client.go:
```diff
@@ -366,10 +393,31 @@ func (c *BitcoindClient) RescanBlocks(
+	for hash, headerRequest := range request {
+		hash := hash
+
+		header, err := headerRequest.Receive()
```
Could this function also use `GetBlocksBatch` below?

bhandras commented on chain/bitcoind_conn.go:
```diff
@@ -425,6 +436,73 @@ func (c *BitcoindConn) GetBlock(hash *chainhash.Hash) (*wire.MsgBlock, error) {
+func (c *BitcoindConn) GetBlocksBatch(
+	hashes []*chainhash.Hash) ([]*wire.MsgBlock, error) {
+
+	batchRequest := make(
```
nit: formatting

bhandras commented on chain/bitcoind_client.go:
```diff
@@ -12,6 +12,7 @@ import (
 	"github.com/btcsuite/btcd/btcjson"
 	"github.com/btcsuite/btcd/btcutil"
 	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/rpcclient"
```
nit: commit messages should be formatted such that the first line includes a short message then a line is kept empty, and finally you add a longer explanation. In this case something like:

```
chain: enable batched operations with bitcoind

In this commit we add support for batched operations
when using the bitcoind client.
```

bhandras replied:
nit: I'd rephrase your commit message to:

```
chain: add batched async calls support for bitcoind client

This commit adds a batch client connection and async RPC
calls to the bitcoind client which will enable batching multiple
requests speeding up operations by reducing round trips.
```

bhandras commented on chain/bitcoind_conn.go:
```diff
@@ -443,7 +443,9 @@ func (c *BitcoindConn) GetBlocksBatch(
 
 	batchRequest := make(
-		map[*chainhash.Hash]rpcclient.FutureGetBlockResult, len(hashes))
+		map[*chainhash.Hash]rpcclient.FutureGetBlockResult,
```
Changes to `GetBlocksBatch` should be in the previous commit.

bhandras commented on chain/bitcoind_conn.go:
```diff
@@ -469,8 +471,7 @@ func (c *BitcoindConn) GetBlocksBatch(
 		if !isBlockPrunedErr(err) || c.prunedBlockDispatcher == nil {
-			batchRawBlocks = append(batchRawBlocks, nil)
-			continue
+			return nil, err
```
Make sure to also update the above comment (in the previous commit once you squash these changes).

Vib-UX committed:
chain: adds GetBlocksBatch() to the bitcoind client

This will leverage the updated batch client to fetch multiple blocks in
one go which can speed up rescan and filter blocks.

Vib-UX committed:
chain: adds RescanBlocksBatched() and FilterBlocksBatched()

Both RescanBlocksBatched() and FilterBlocksBatched() leverages GetBlocksBatch()
which works with the updated batch client for speed ups.

bhandras commented on chain/bitcoind_conn.go:
```diff
@@ -94,6 +94,9 @@ type BitcoindConn struct {
 	// client is the RPC client to the bitcoind node.
 	client *rpcclient.Client
 
+	// batchClient is the RPC client to the bitcoind node.
```
nit: 
```
// batchClient is the batched RPC client to the bitcoind node.
```

bhandras commented on chain/bitcoind_conn.go:
```diff
@@ -234,6 +243,8 @@ func (c *BitcoindConn) Stop() {
 		log.Errorf("error shutting down bitcoind events: %w", err)
 	}
 
+	c.batchClient.Shutdown()
```
Do we also need `c.batchClient.WaitForShutdown()` like we do for the normal client (line 252).

bhandras commented on chain/bitcoind_client.go:
```diff
@@ -160,6 +160,13 @@ func (c *BitcoindClient) GetBlock(hash *chainhash.Hash) (*wire.MsgBlock, error)
 	return c.chainConn.GetBlock(hash)
 }
 
+// GetBlocksBatch returns batchBlocks from batchHashes.
```
nit: commit message:
```
chain: add GetBlocksBatch() to the bitcoind client

This will leverage the updated batch client to fetch multiple blocks in
one go which can speed up rescan and filter blocks.
```

bhandras commented on chain/bitcoind_client.go:
```diff
@@ -427,10 +427,44 @@ func (c *BitcoindClient) RescanBlocks(
 	blockHashes []chainhash.Hash) ([]btcjson.RescannedBlock, error) {
 
 	rescannedBlocks := make([]btcjson.RescannedBlock, 0, len(blockHashes))
+
```
nit: in commit message:
```
chain: updates RescanBlocks() & FilterBlocks()

Both RescanBlocks() and FilterBlocks() now leverages GetBlocksBatch()
which works with the updated batch client for speed ups.
```

bhandras commented on chain/bitcoind_client.go:
```diff
@@ -427,10 +427,44 @@ func (c *BitcoindClient) RescanBlocks(
+		headerRequest, err := c.GetBlockHeaderVerboseAsync(&hash)
+		if err != nil {
+			log.Warnf(
+				"Unable to receive header promise from hash %v: %s", hash, err)
```
nit: format to 80 columns.

bhandras commented on chain/bitcoind_client.go:
```diff
@@ -427,10 +427,44 @@ func (c *BitcoindClient) RescanBlocks(
+		requests[hash] = headerRequest
+	}
+
+	c.SendAsyncQueue()
```
I think this function could be reworked a bit from its current form, to:

1) fetch block headers to get timestamps and heights (send async and receive too)
2) in the receive loop you could filter out those which are not important for us:
```
if !c.shouldFilterBlock(time.Unix(header.Time, 0)) {
			continue
}
```
3) finally get the filtered block hashes

bhandras commented on chain/bitcoind_client.go:
```diff
@@ -981,17 +1008,21 @@ func (c *BitcoindClient) FilterBlocks(
 
 	blockFilterer := NewBlockFilterer(c.chainConn.cfg.ChainParams, req)
 
+	storeHash := make([]*chainhash.Hash, 0, len(req.Blocks))
```
nit: rename `storeHash` to `blockHashes`

bhandras commented on chain/bitcoind_client.go:
```diff
@@ -981,17 +1008,21 @@ func (c *BitcoindClient) FilterBlocks(
 
+	storeHash := make([]*chainhash.Hash, 0, len(req.Blocks))
+
+	for _, block := range req.Blocks {
```
nit: better to use the form: `for i := range req.Blocks` and then refer to the i'th block as `req.Blocks[i]`.

bhandras commented on chain/bitcoind_client.go:
```diff
@@ -981,17 +1008,21 @@ func (c *BitcoindClient) FilterBlocks(
-			return nil, err
-		}
-
+	for i, rawBlock := range rawBlocks {
```
nit: same here, would not surface `rawBlock` to avoid copy.

bhandras commented on chain/bitcoind_client.go:
```diff
@@ -1003,14 +1034,15 @@ func (c *BitcoindClient) FilterBlocks(
 		}
 
 		return resp, nil
+
```
nit: extra nl

bhandras reviewed (changes requested):
Great work! Getting very close. `RescanBlocks()` needs some more fixing but other than that almost good to go.

Vib-UX committed:
chain: testcoverage for GetBlocksBatch()

Testcoverage to ensure that GetBlocksBatch() which leverages batchAPI
works as expected with enhanced performance.

Vib-UX committed:
chain: testcoverage for RescanBlocksBatched()

Testcoverage to ensure that RescanBlocksBatched() which leverages
updated batchAPI for scanning the chain works as expected with enhanced
performance.

Vib-UX committed:
chain: adds utility functions for testing

Vib-UX committed:
chain: testcoverage for FilterBlocksBatched()

Testcoverage to ensure that FilterBlocksBatched() which leverages
batchAPI for scanning the chain works as expected with enhanced
performance.

bhandras commented on chain/bitcoind_client.go:
```diff
@@ -426,14 +426,38 @@ func (c *BitcoindClient) LoadTxFilter(reset bool, filters ...interface{}) error
 		if err != nil {
 			log.Warnf("Unable to get header %s from bitcoind: %s",
-				hash, err)
+				blockHashes[i], err)
```
nit: can just use `hash` here.

bhandras commented on chain/bitcoind_client.go:
```diff
@@ -442,27 +466,53 @@ func (c *BitcoindClient) RescanBlocks(
+			return nil, err
 		}
+		batchedBlockHashPointers := blockHashPointers[start:end]
+		index := -1
```
nit: not sure what the role of the variable `index` is?

Vib-UX replied:
Yes, this is redundant now since I am traversing using index, it was there as per earlier implementation which used maps for traversal. Will fix it thanks for pointing it out!

bhandras commented on chain/bitcoind_client.go:
```diff
@@ -442,27 +466,53 @@ func (c *BitcoindClient) RescanBlocks(
+					Hash: hash.String(),
+				}
+				for _, tx := range relevantTxs {
+					rescannedBlock.Transactions = append(
```
nit: `rescannedBlock.Transactions` can be preallocated.

bhandras commented on chain/bitcoind_client.go:
```diff
@@ -981,36 +1031,54 @@ func (c *BitcoindClient) FilterBlocks(
+			// The `BatchIndex` is returned so that the caller can
+			// compute the *next* block from which to begin again.
+			resp := &FilterBlocksResponse{
+				BatchIndex:         uint32(j + start),
```
nit: maybe `start + j` is more logical here since `start` is the offset, and `j` is the index in the batch.

bhandras commented on chain/bitcoind_client_test.go:
```diff
@@ -0,0 +1,127 @@
+	timeout := time.After(10 * time.Second)
+	for {
+		_, bitcoindHeight, err := bitcoindClient.GetBestBlock()
+		if err != nil {
```
nit: can use `require.NoError` here too.

bhandras commented on chain/bitcoind_client_test.go:
```diff
@@ -0,0 +1,127 @@
+		select {
+		case <-time.After(100 * time.Millisecond):
+		case <-timeout:
+			t.Fatalf("timed out waiting to sync notifier")
```
nit: timeout message should be updated.

bhandras commented on chain/bitcoind_client_test.go:
```diff
@@ -0,0 +1,127 @@
+		require.NoError(t, err)
+		msgBlocksStandard = append(msgBlocksStandard, msgBlock)
+	}
+	msgBlocksBatchClient, err := bitcoindClient.GetBlocksBatch(blockHashes)
```
Nice test!

bhandras commented on chain/bitcoind_client_test.go:
```diff
@@ -125,3 +125,52 @@ func testBitcoindGetBlocksBatch(t *testing.T, rpcpolling bool) {
+// TestBitcoindRescanBlocks ensures that we correctly retrieve the required details
+// for the watched txn.
+func TestBitcoindRescanBlocks(t *testing.T) {
+	testBitcoindRescanBlocks(t, true)
```
nit: can create a subtest here with:

`t.Run(fmt.Sprintf("rpcpoll=true"), func(t *testing.T) { ... })` etc

bhandras commented on chain/bitcoind_client_test.go:
```diff
@@ -125,3 +125,52 @@ func testBitcoindGetBlocksBatch(t *testing.T, rpcpolling bool) {
+	testBitcoindRescanBlocks(t, true)
+	testBitcoindRescanBlocks(t, false)
+}
+func testBitcoindRescanBlocks(t *testing.T, rpcpolling bool) {
```
nit: add extra nl above

bhandras commented on chain/bitcoind_client_test.go:
```diff
@@ -174,3 +176,78 @@ func testBitcoindRescanBlocks(t *testing.T, rpcpolling bool) {
+	testBitcoindFilterBlocks(t, true)
+	testBitcoindFilterBlocks(t, false)
+}
+func testBitcoindFilterBlocks(t *testing.T, rpcpolling bool) {
```
nit: missing extra nl

bhandras reviewed (approved):
LGTM, just a few nits remain, great work @Vib-UX :tada:

bhandras commented:
Actually one last minute comment that maybe needs more work: it is a bit unfortunate that we need to add back old code in order to test expectations, perhaps a better way would be to instead check the output of the new batched call since we know what the output should look like. Otherwise there'll always remain some unused code as part of the public interface that we'll never use.

Alternatively you can also rename the "Old" functions as "Sync", meaning they're non batched, slow sequential calls. Also what do you think @guggero ?

Vib-UX commented:
> Actually one last minute comment that maybe needs more work: it is a bit unfortunate that we need to add back old code in order to test expectations, perhaps a better way would be to instead check the output of the new batched call since we know what the output should look like. Otherwise there'll always remain some unused code as part of the public interface that we'll never use.
> 
> Alternatively you can also rename the "Old" functions as "Sync", meaning they're non batched, slow sequential calls. Also what do you think @guggero ?

<img width="464" alt="btcwallet-testCoverageSpeedUp" src="https://user-images.githubusercontent.com/40585900/182925696-54e6163e-3bcf-4f4c-99b6-0dc09cbc3f85.png">

The updated testcoverage removes dependency from the previous Sync functions making it redundant. It was earlier there to compare speedup.

guggero commented:
> Alternatively you can also rename the "Old" functions as "Sync", meaning they're non batched, slow sequential calls. Also what do you think @guggero ?

Just jumping in to answer the question, I'm probably lacking most of the context of the previous discussion. But I'd generally not rename any existing functions, so other users (`lnd` isn't the only software that uses `btcwallet` AFAIK) don't need to change any code when updating. And using the same name as before for something that works differently is not a good idea either. So I'd keep everything existing as it is and add `*Async` methods for new stuff (which I believe is what's currently being done?).

bhandras commented:
> > Alternatively you can also rename the "Old" functions as "Sync", meaning they're non batched, slow sequential calls. Also what do you think @guggero ?
> 
> Just jumping in to answer the question, I'm probably lacking most of the context of the previous discussion. But I'd generally not rename any existing functions, so other users (`lnd` isn't the only software that uses `btcwallet` AFAIK) don't need to change any code when updating. And using the same name as before for something that works differently is not a good idea either. So I'd keep everything existing as it is and add `*Async` methods for new stuff (which I believe is what's currently being done?).

Yeah maybe best to keep existing functionality and unrenamed. However since the new functionality is not exactly the same semantics as when using the "Async" versions of the RPC API, we could name them "Batched" perhaps to refer that it is doing batched calls under the hood.

Vib-UX commented:
rebase master : [`a6ba11e`](https://github.com/Vib-UX/btcwallet/commit/a6ba11e2d983bc556218c65703da656367975111) to [`36bcc1e`](https://github.com/Vib-UX/btcwallet/commit/36bcc1ea2655b636937d31c7223d8c09d3569c45)

bhandras commented on chain/bitcoind_conn.go:
```diff
@@ -437,6 +437,73 @@ func (c *BitcoindConn) GetBlock(hash *chainhash.Hash) (*wire.MsgBlock, error) {
+func (c *BitcoindConn) GetBlocksBatch(
+	hashes []*chainhash.Hash) ([]*wire.MsgBlock, error) {
+
+	batchRequest := make(
```
nit: `batchRequest` could actually just be a slice, no need to key it with the hashes since we always iterate on `hashes`.

bhandras commented on chain/bitcoind_client.go:
```diff
@@ -469,6 +469,103 @@ func (c *BitcoindClient) RescanBlocks(
+
+	// requests map will store rpcclient.FutureGetBlockHeaderVerboseResult
+	// for each blockHash.
+	requests := make(
```
nit: i think this also could just be an array since we always iterate on `blockHashes`.

bhandras commented on chain/bitcoind_client.go:
```diff
@@ -1017,6 +1114,71 @@ func (c *BitcoindClient) FilterBlocks(
+// a FilterBlocksReponse for the first block containing a matching address.
+// If no matches are found in the range of blocks requested, the returned
+// response will be nil.
+
```
nit: extra nl

bhandras commented on chain/bitcoind_client_test.go:
```diff
@@ -0,0 +1,126 @@
+		require.NoError(t, err)
+		msgBlocksStandard = append(msgBlocksStandard, msgBlock)
+	}
+	msgBlocksBatchClient, err := bitcoindClient.GetBlocksBatch(blockHashes)
```
nit: add an extra nl after the for block

bhandras commented on chain/bitcoind_client_test.go:
```diff
@@ -0,0 +1,126 @@
+	// linearly with GetBlock() (store the result for each GetBlock() in a
+	// slice) further also try to retrieve blocks in batch with GetBlocksBatch().
+	// At last will compare the results from both which should be equal.
+	if _, err := miner.Client.Generate(5); err != nil {
```
nit: I think it'd be better to generate something like 2x15 + some blocks to test that we do create multiple batches correctly.

bhandras reviewed (approved):
LGTM, just a few nits remain, great work! :tada:  Make sure to add a detailed description when you open the pull request against the main repo! :rocket:

guggero commented:
I think it makes sense to open the PR against the main repo now. I'll happily do a review there. Is that okay for everyone?

bhandras commented:
> I think it makes sense to open the PR against the main repo now. I'll happily do a review there. Is that okay for everyone?

Yeah, sure. Actually we also just agree with @Vib-UX to open it against upstream, since the PR is quite ready as is.

Vib-UX committed:
chain: refactors and adds generic fnc()
